Command Palette: Add site switcher command

dsas opened the pull request:
Fixes DOTCOM-15345

## Proposed changes:
* Add a new site switcher feature to the WordPress Command Palette (requires WP 6.9+)
* Implement REST API proxy endpoint at `/jetpack/v4/sites/compact` to fetch user's WordPress.com sites via the user's Jetpack connection
* Add client-side localStorage caching (1 hour duration) with user-specific cache keys to minimize API requests
* Implement search filtering to quickly find sites by name or domain
* Display site icons with fallback to generic site logo
* Support dynamic command labels based on whether site has a name or just domain

### Implementation details:
* `_inc/site-switcher-endpoint.php`: REST endpoint for Jetpack sites that proxies the public-api endpoint
* `_inc/site-switcher.php`: JS enqueueing, and config
* `_inc/site-switcher.js`: React hooks-based command loader with search filtering and caching
* Commands show as "Switch to Site Name (domain.com)" or "Switch to: domain.com" if no name
* Generic search terms like "site" or "switch site" show all sites; specific searches filter results

### Other information:

- [x] Have you written new tests for your changes, if applicable? - No tests added (following pattern of other `_inc/*.js` files)
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion


## Does this pull request change what data or activity we track or use?
Not really. This feature only fetches existing site data through the established Jetpack connection API.

## Testing instructions:

### Prerequisites:
* Jetpack plugin installed and connected to WordPress.com
* User must have multiple WordPress.com sites

### Testing steps:
1. Install and activate Jetpack plugin from this branch
On Simple you need to also apply 198424-ghe-Automattic/wpcom
2. Ensure your site is connected to WordPress.com with a user account that has multiple sites
3. On a Jetpack site navigate to any wp-admin page, on a simple or WoA site then navigate to the site-editor (core command palette is blocked by wp command palette otherwise)
4. Open the Command Palette:
   - Mac: `Cmd + K`
   - Windows/Linux: `Ctrl + K`
5. Type "site" or "switch site" - you should see all your WordPress.com sites listed
6. Try searching for a specific site name or domain - results should filter accordingly
7. Click on a site command - you should be redirected to that site's wp-admin

### Expected behavior:
* All sites show on search for generic keywords
* Search filters work for both site names and domains
* Icons display correctly (site-specific or fallback logo)
* Navigation to selected site works
* Subsequent page loads use cached data (check Network tab - no `compact` request for 1 hour)

### Cache testing:
1. Open browser DevTools > Network tab
2. Open command palette and search for sites - you should see a request to `/jetpack/v4/sites/compact`
3. Reload the page and open command palette again - no new request (using localStorage cache)
4. Wait 1 hour or manually clear localStorage to see cache refresh

https://github.com/user-attachments/assets/14cb805f-0f21-4f03-af5a-0f4973ef445b

dsas committed:
Add switch site command to the command palette

This command allows the user to search for and switch any of their
WordPress.com connected sites.

dsas commented on projects/plugins/jetpack/_inc/site-switcher.php:
```diff
@@ -0,0 +1,106 @@
+/**
+ * Register REST API endpoint to fetch compact sites list from WordPress.com
+ */
+function jetpack_site_switcher_register_rest_routes() {
```
Needs moving to a v2 endpoint

dsas committed:
Change the API endpoint

dsas committed:
Change API endpoint again

dsas committed:
api

dsas commented on projects/plugins/jetpack/_inc/site-switcher.js:
```diff
@@ -0,0 +1,218 @@
+			const domain = site.URL ? new URL( site.URL ).hostname : '';
+
+			const iconElement = site.icon?.img ? (
+				<img src={ site.icon.img } alt={ site.name } />
```
seems far-fetched, there are bigger problems in that case

dsas commented on projects/plugins/jetpack/_inc/site-switcher.js:
```diff
@@ -0,0 +1,218 @@
+function useSiteSwitcherCommandLoader( { search } ) {
+	const [ sites, setSites ] = useState( [] );
+	const [ isLoading, setIsLoading ] = useState( true );
+
+	// Fetch sites on mount
+	useEffect( () => {
+		fetchSitesFromWordPressCom()
+			.then( fetchedSites => {
+				setSites( fetchedSites );
+				setIsLoading( false );
+			} )
+			.catch( () => {
+				setIsLoading( false );
+			} );
+	}, [] );
```
I think reloading the page is fine - the commands are to navigate away from the current page anyway

dsas commented on projects/plugins/jetpack/_inc/site-switcher.js:
```diff
@@ -0,0 +1,218 @@
+				label,
+				icon: iconElement,
+				callback: ( { close } ) => {
+					window.location.href = `${ site.URL }/wp-admin`;
```
The api is the only place we can get a list of valid domains from

dsas commented on projects/plugins/jetpack/_inc/site-switcher.js:
```diff
@@ -0,0 +1,212 @@
+import { sprintf, __ } from '@wordpress/i18n';
+import { siteLogo } from '@wordpress/icons';
+
+const userId = window.jetpackSiteSwitcherConfig?.userId || 0;
+const CACHE_KEY = `jetpack_site_switcher_sites_${ userId }`;
```
the cache is stored in local storage, so all of these users would need to use the same browser on the same computer. they will be logged in because the functionality is just for logged in users

dsas committed:
address review comments

dsas commented on projects/plugins/jetpack/_inc/site-switcher.js:
```diff
@@ -0,0 +1,212 @@
+		const isGenericSearch =
+			! searchLower ||
+			searchLower === __( 'site', 'jetpack' ).toLowerCase() ||
+			searchLower === __( 'switch site', 'jetpack' ).toLowerCase();
```
Other built-in commands work this way.

dsas commented on projects/plugins/jetpack/_inc/site-switcher.js:
```diff
@@ -0,0 +1,218 @@
+			const domain = site.URL ? new URL( site.URL ).hostname : '';
+
+			const iconElement = site.icon?.img ? (
+				<img src={ site.icon.img } alt={ site.name } />
```
It's a decorative image really, a blank alt text is more appropriate.

dsas committed:
address review comments

dsas committed:
Loosen availability check

Command palette introduced in WP 6.3.

dsas committed:
Exclude current site from switcher

dsas committed:
Make unconnected user experience more robust

dsas committed:
Add switch keyword after feedback from testers

rcrdortiz reviewed (dismissed):
Working as expected for me.

dsas committed:
Improve handling of generic search words

Let people search for 'switch site dean' to limit the returned sites to
those matching dean, rather than those matching 'switch site dean'.

dsas commented on projects/plugins/jetpack/_inc/site-switcher.js:
```diff
@@ -0,0 +1,259 @@
+		const currentURL = untrailingslashit( window.location.href.toLowerCase() );
+		const otherSites = filteredSites.filter( site => {
+			if ( ! site.URL ) {
+				return true;
+			}
+			// Normalize site URL for comparison
+			const siteURL = untrailingslashit( site.URL.toLowerCase() );
+			// Check if current URL starts with site URL (handles multisite subdirectory installs)
+			// e.g., current: example.com/site1/wp-admin matches site: example.com/site1
+			return ! currentURL.startsWith( siteURL );
```
this is why it uses startswith

dsas commented on projects/plugins/jetpack/_inc/site-switcher.js:
```diff
@@ -0,0 +1,259 @@
+		return cachedSites;
+	}
+
+	const apiPath = window.jetpackSiteSwitcherConfig?.apiPath;
+
+	try {
+		const data = await apiFetch( {
+			path: apiPath,
+			method: 'GET',
+			global: true,
+		} );
```
it seems very very unlikely and it's covered by the try...catch.

dsas commented on projects/plugins/jetpack/_inc/site-switcher.js:
```diff
@@ -0,0 +1,259 @@
+
+		return otherSites.map( site => {
+			// Extract domain from URL for display - don't want to display the protocol.
+			const domain = getHostnameFromURL( site.URL );
+
+			const iconElement = site.icon?.img ? <img src={ site.icon.img } alt="" /> : siteLogo;
+
+			// Use site name if available, otherwise just show domain
+			const label = site.name
+				? sprintf(
+						/* translators: %1$s: site name, %2$s: site domain */
+						__( 'Switch to %1$s (%2$s)', 'jetpack' ),
+						site.name,
+						domain
+				  )
+				: sprintf(
+						/* translators: %s: site domain */
+						__( 'Switch to %s', 'jetpack' ),
+						domain
+				  );
```
Seems impossible, and i'm not sure what a good fallback label would be.

dsas commented on projects/plugins/jetpack/_inc/site-switcher.js:
```diff
@@ -0,0 +1,259 @@
+			// Extract domain from URL for display - don't want to display the protocol.
+			const domain = getHostnameFromURL( site.URL );
+
+			const iconElement = site.icon?.img ? <img src={ site.icon.img } alt="" /> : siteLogo;
```
The API is trusted, react should take care of any escaping issues.

dsas commented on projects/plugins/jetpack/_inc/site-switcher.php:
```diff
@@ -0,0 +1,75 @@
+ * Enqueue site switcher scripts in all wp-admin pages.
+ */
+function jetpack_site_switcher_enqueue_scripts() {
+	// Require WordPress 6.3+ for Command Palette support
+	// The command palette is available in the block editor since 6.3, and admin-wide since 6.9
+	if ( version_compare( get_bloginfo( 'version' ), '6.3-alpha', '<' ) ) {
```
AI getting confused. It requires WP6.9 to function properly in any old wp-admin page, but is available from 6.3 onwards in the site editor.

dsas committed:
escape keyword in regex

dsas committed:
Filter out sites with unparseable urls

dsas committed:
Add wp-icons to explicit dependencies

heavyweight reviewed (approved):
Followed testing steps and LGTM
